### PR TITLE
Add handler for additional post messages from interactive examples, and a refactor

### PR DESCRIPTION
**Pull request by schalkneethling**
@tkadlec r? This is the other side of https://github.com/mdn/interactive-examples/pull/977 - I also did some refactoring to make the code more modular and easier to understand/read.

@jwhitlock Opening this for code review by Tim, I first need to have https://github.com/mdn/interactive-examples/pull/977 reviewed and merged so that I can test this properly. So, will mark as WIP for now.

### CODE REVIEW

File: kuma/static/js/utils/utils.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+window.mdn.utils = {
+    /**
+     * Generate and returns a random string thanks to:
+     * https://stackoverflow.com/questions/1349404/generate-random-string-characters-in-javascript
+     * @param {Number} strLength - The length of the string to return
+     * @return {String} a randomly generated string with a chracter count of `strLength`
+     */
+    randomString: function(strLength) {
```
Comment (jwhitlock) [inline kuma/static/js/utils/utils.js]: This refactored code works in the console. ``randomString()`` now returns an empty string, but all the calls have been updated to ``mdn.utils.randomString(5)``.  I have a slight preference for a default, like ``strLength = Math.min(5, strLength)``.

File: kuma/static/js/utils/post-message-handler.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        if (event.origin !== allowedOrigin) {
+            return false;
+        }
+
+        console.info('====== eventData.label ======', eventData.label);
+
+        if (eventData.label === 'Performance Events') {
+            handlePerformanceEvents(eventData);
+        } else if (eventData.measureName) {
+            handlePerfMarks(eventData);
```
Comment (jwhitlock) [inline kuma/static/js/utils/post-message-handler.js]: This appears to be the new functionality.

File: kuma/static/js/utils/post-message-handler.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+/**
+ * Processes marks and sends beacons to GA based on mark value
+ * @param {Object} perfData - Object containing performance mark/measure information
+ */
+function handlePerfMarks(perfData) {
+    // if there is no `measureName` property, just set a mark
+    if (!perfData.measureName) {
+        window.mdn.perf.setMark(perfData.markName);
```
Comment (jwhitlock) [inline kuma/static/js/utils/post-message-handler.js]: Based on current logic, we'll never run this branch - it is only called when ``measureName`` is defined.

File: kuma/static/js/utils/post-message-handler.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    } else {
+        window.mdn.perf.setMark(perfData.markName);
+        window.mdn.perf.setMeasure(
```
Comment (jwhitlock) [inline kuma/static/js/utils/post-message-handler.js]: This doesn't match the definition of ``setMeasure``, which is expecting a single ``measureData`` value. I have a slight preference for changing ``setMeasure`` to expect these three variables.

File: kuma/static/js/utils/utils.js (hunk 1 above)
Comment (schalkneethling) [reply kuma/static/js/utils/utils.js]: Yeah, I agree. My initial thought was to provide a default if nothing is passed, I should do that.

File: kuma/static/js/utils/post-message-handler.js (hunk 3 above)
Comment (schalkneethling) [reply kuma/static/js/utils/post-message-handler.js]: Doh!, that should be `markName`

File: kuma/static/js/utils/post-message-handler.js (hunk 4 above)
Comment (schalkneethling) [reply kuma/static/js/utils/post-message-handler.js]: Doh! again. This is why I specifically mentioned that this should not be merged and is largely untested code :) It should be passing an object of course. Thanks @jwhitlock

**Comment by schalkneethling** [comment]
@jwhitlock This has been updated, thanks for the review. This should still not be merged until https://github.com/mdn/interactive-examples/pull/977 has been, and I have had a chance to do some testing locally.

I have also updated the commit message to include the bug number reference.

**Comment by schalkneethling** [comment]
Alrighty, I have now tested this locally, and data is being sent to GA as expected for all editors, see below:

## Tabbed Editor

tabbed-editor

## CSS Editor

css-editor

## JS Editor

js-editor

**Comment by schalkneethling** [comment]
Also, User Timing can be seen collected in devtools:

screen shot 2018-06-26 at 13 56 58

With that I am calling this ready to merge and push, baring any additional review comments.

File: kuma/static/js/utils/post-message-handler.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        window.mdn.perf.setMeasure({
+            measureName: perfData.measureName,
+            startMark: perfData.startMark,
+            endMark: perfData.endMark
+        });
+
+        mdn.analytics.trackTiming({
+            category: 'RUM - Interactive Examples',
+            timingVar: perfData.measureName,
+            value: window.mdn.perf.getDuration(perfData.measureName)
```
Comment (jwhitlock) [inline kuma/static/js/utils/post-message-handler.js]: The [type of the timing event is an integer](https://developers.google.com/analytics/devguides/collection/analyticsjs/field-reference#timingValue), but this is passing a float. We should truncate it on our side, like ``Math.round(window.mdn.perf.getDuration(perfData.measureName))``, or adjust ``get_duration`` to return an integer.

File: kuma/static/js/utils/post-message-handler.js (hunk 3 above)
Comment (jwhitlock) [reply kuma/static/js/utils/post-message-handler.js]: 👍 I see ``interactiveExamplesMsgHandler`` changed to check for ``markName``.

File: kuma/static/js/utils/post-message-handler.js (hunk 5 above)
Comment (schalkneethling) [reply kuma/static/js/utils/post-message-handler.js]: Aha, good catch. I did not know it had this restriction/requirement. We can simply ensure that we never return a floating point in a central location.

**Comment by schalkneethling** [comment]
@jwhitlock So, the `ie-time-*` values are from inside the `iframe` itself. `Time to iframe fetch start` is from outside i.e. the parent page.

**Comment by schalkneethling** [comment]
@jwhitlock Updated. Thanks!

**Comment by jwhitlock** [comment]
I thought if I adjusted the subject line of the merge commit, then [bug 143](https://bugzilla.mozilla.org/show_bug.cgi?id=143) wouldn't be updated. I guess I needed to remove more than just the hash ☹️ . Try to omit "issue 143" from the next branch name.

**Pull request merged**